File: app/src/main/java/com/example/chris/atp_music_player/ui/AlbumListFragment.java
package com.example.chris.atp_music_player.ui;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.example.chris.atp_music_player.R;
import com.example.chris.atp_music_player.adapters.AlbumListAdapter;
import com.example.chris.atp_music_player.models.Album;
import com.example.chris.atp_music_player.provider.MusicProvider;

import java.util.List;

import butterknife.ButterKnife;
import butterknife.InjectView;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

public class AlbumListFragment extends Fragment {
    @InjectView(R.id.artist_album_rv)
    RecyclerView mRecyclerView;
    @InjectView(R.id.error_placeholder)
    TextView mErrorView;

    public static AlbumListFragment newInstance() {
        return new AlbumListFragment();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.artist_album_list, container, false);
        ButterKnife.inject(this, view);

        mRecyclerView.setHasFixedSize(true);
        mRecyclerView.setLayoutManager(new GridLayoutManager(getActivity(), 2));

        loadAlbums();
        return view;
    }

    private void loadAlbums() {
        MusicProvider musicProvider = new MusicProvider(getActivity());
        musicProvider.getAllAlbums().subscribeOn(Schedulers.newThread()).observeOn(AndroidSchedulers.mainThread()).subscribe(new Subscriber<List<Album>>() {
            @Override
            public void onCompleted() {
            }

            @Override
            public void onError(Throwable e) {
            }

            @Override
            public void onNext(List<Album> albums) {
                if (albums.isEmpty()) {
                    mErrorView.setText(R.string.err_albums);
                    mRecyclerView.setVisibility(View.GONE);
                    mErrorView.setVisibility(View.VISIBLE);
                } else {

                    mRecyclerView.setAdapter(new AlbumListAdapter(getActivity(), albums));
                }
            }
        });
    }
}

File: app/src/main/java/com/example/chris/atp_music_player/ui/AlbumListFragment.kt
package com.example.chris.atp_music_player.ui

import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.GridLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.example.chris.atp_music_player.ATPApplication
import com.example.chris.atp_music_player.Constants
import com.example.chris.atp_music_player.R
import com.example.chris.atp_music_player.adapters.AlbumListAdapter
import com.example.chris.atp_music_player.playback.MediaProvider
import kotlinx.android.synthetic.main.artist_album_list.*
import kotlinx.android.synthetic.main.no_data.*
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

class AlbumListFragment : Fragment() {
    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater?.inflate(R.layout.artist_album_list, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        artist_album_rv.setHasFixedSize(true)
        artist_album_rv.layoutManager = GridLayoutManager(context, 2)

        loadAlbums()
    }

    private fun loadAlbums() {
        MediaProvider(context)
                .getAllAlbums()
                .toList()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ albums ->
                    if (albums.isEmpty()) {
                        showErrorView()
                    } else {
                        artist_album_rv.adapter = AlbumListAdapter(albums, { album ->
                            val intent = Intent(context, SongSubsetActivity::class.java)
                            intent.putExtra(Constants.QUERY_CONSTRAINT, album.title)
                            intent.putExtra(Constants.QUERY_TYPE, Constants.QUERY_TYPE_ALBUM)
                            intent.putExtra(Constants.DATA_ALBUM_ID, album.id)
                            ATPApplication.subActivityWillBeVisible()
                            context.startActivity(intent)
                        })
                    }
                }, {
                    showErrorView()
                })
    }

    private fun showErrorView() {
        error.setText(R.string.err_albums)
        error.visibility = View.VISIBLE
        artist_album_rv.visibility = View.GONE
    }
}
